lighthouse-plugin-ad-speed-insights/gatherers/iframe-elements.js

```diff
@@ -0,0 +1,74 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// @ts-ignore
+const Gatherer = require('lighthouse/lighthouse-core/gather/gatherers/gatherer.js');
+// @ts-ignore
+const pageFunctions = require('lighthouse/lighthouse-core/lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ * TODO(jburger): Move to page-functions.js once integrated with Lighthouse.
+ * @param {HTMLIFrameElement} element
+ * @return {DOMRect | ClientRect}
+ */
+function getBoxModel(element) {
+  return element.getBoundingClientRect();
+}
+/**
+ * @return {Array<LH.Crdp.DOM.BoxModel>}
+ */
+function collectIframeElements() {
+  // @ts-ignore - put into scope via stringification
+  const iframeElements = getElementsInDocument('iframe'); // eslint-disable-line no-undef
+  return iframeElements.map(/** @param {HTMLIFrameElement} node */ (node) => {
+    // @ts-ignore - put into scope via stringification
+    const outerHTML = getOuterHTMLSnippet(node); // eslint-disable-line no-undef
+    // @ts-ignore
+    const boxModel = getBoxModel(node).toJSON();
```

Comment:
https://github.com/GoogleChrome/lighthouse/blob/520177b74202ac2ac5ca7e2f6daa2a7cc6bd5612/lighthouse-core/gather/gatherers/image-elements.js#L25 manually copies the properties because ClientRect does not JSONify, but document.querySelector('body').getBoundingClientRect().toJSON() seemed to work fine, so I think this is fine, just noting the difference.

Reply:
> manually copies the properties because ClientRect does not JSONify

I really couldn't remember why that was in there because it does run fine without it, but it turns out since that code was written in early 2017, [`getBoundingClientRect()` was changed to return `DOMRect` instead of `ClientRect`](https://chromium-review.googlesource.com/559978), so now it is JSONifiable.

This code can only run in Chrome from over the protocol, so seems totally fine to rely on that behavior (we might clean that up in `ImageElements` now :)

Reply:
> we might clean that up in ImageElements now

already in progress :) https://github.com/GoogleChrome/lighthouse/pull/7785

Reply:
@brendankenny Once we merge this it might make sense to move getClientRect() to pageFunctions for both gatherers to use.

While it is JSONifiable, typescript doesn't seem to think so, hence the `@ts-ignore`.

Reply:
> While it is JSONifiable, typescript doesn't seem to think so, hence the @ts-ignore.

it's because it's `DOMRect | ClientRect` (since you still get a `ClientRect` in at least IE, and the types have to try to account for all browsers)